src/Compilers/CSharp/Test/Syntax/Parsing/ParsingTests.cs

```diff
@@ -145,28 +171,34 @@ protected CSharpSyntaxNode UsingNode(string text, CSharpParseOptions options, pa
         /// </summary>
         protected void UsingNode(CSharpSyntaxNode root)
         {
-            var nodes = EnumerateNodes(root);
-#if PARSING_TESTS_DUMP
-            nodes = nodes.ToArray(); //force eval to dump contents
-#endif
+            VerifyEnumeratorConsumed();
+            _node = root;
+            var nodes = EnumerateNodes(root, dump: false);
             _treeEnumerator = nodes.GetEnumerator();
         }
 
         /// <summary>
         /// Moves the enumerator and asserts that the current node is of the given kind.
         /// </summary>
         [DebuggerHidden]
-        protected SyntaxNodeOrToken N(SyntaxKind kind, string value = null)
+        protected SyntaxNodeOrToken N(SyntaxKind kind, string? value = null)
         {
-            Assert.True(_treeEnumerator.MoveNext());
-            Assert.Equal(kind, _treeEnumerator.Current.Kind());
+            try
+            {
+                Assert.True(_treeEnumerator!.MoveNext());
+                Assert.Equal(kind, _treeEnumerator.Current.Kind());
+
+                if (value != null)
+                {
+                    Assert.Equal(_treeEnumerator.Current.ToString(), value);
+                }
 
-            if (value != null)
+                return _treeEnumerator.Current;
+            }
+            catch when (DumpAndCleanup())
```

Comment:
> catch when (DumpAndCleanup()) [](start = 11, length = 30)

assertions now cause to dump. Nice! :)

Reply:
Yes, this was the best way I could figure out to run this cleanup while making the original stack trace show in the xunit output. Bare ‘throw’ didn’t show the original stack in the log for some reason.